agent/consul/gateway_locator_test.go

```diff
@@ -93,17 +106,25 @@ func TestGatewayLocator(t *testing.T) {
 				)
 				g.SetUseReplicationSignal(isLeader)
 
+				t.Run("before first run", func(t *testing.T) {
+					assert.True(t, g.DialPrimaryThroughLocalGateway()) // defaults to sure!
```

Comment:
As someone not super familiar with this code. It's not very clear to me why this test case should expect to dial through gateways when it has no data?

Doesn't that mean it doesn't know where primary gateways are yet and so it should try servers direct?

If this is correct because we just opportunistically dial gateways and fail if there are none a comment here might help!

Reply:
So `DialPrimaryThroughLocalGateway` just keeps track of the simple failure detector logic and does not include information about federation state catalog data.

It primarily gets used this way:
```
		if g.datacenter == g.primaryDatacenter {
			addrs = g.primaryGateways
		} else if g.DialPrimaryThroughLocalGateway() && len(g.localGateways) > 0 {
			addrs = g.localGateways
		} else {
			// Note calling StringSliceMergeSorted only works because both
			// inputs are pre-sorted. If for some reason one of the lists has
			// *duplicates* (which shouldn't happen) it's not great but it
			// won't break anything other than biasing our eventual random
			// choice a little bit.
			addrs = stringslice.MergeSorted(g.primaryGateways, g.PrimaryGatewayFallbackAddresses())
		}
```

In this PR we explicitly initialize this failure detector to the success state on startup:
```
@@ -283,6 +289,8 @@ func NewGatewayLocator(
                primaryGatewaysReadyCh: make(chan struct{}),
        }
        g.logPrimaryDialingMessage(g.DialPrimaryThroughLocalGateway())
+       // initialize
+       g.SetLastFederationStateReplicationError(nil, false)
        return g
 }
```
before actually making an observation. That's the main change.  This means that the gateway locator will just assume that the local gateways are the better option UNTIL they fail, to encourage the system to stick with the local gateways even on startup.  If there aren't actually any local gateways then the calling code will ignore this decision and do something else.